pkg/server/controllers.go

```diff
@@ -410,24 +431,33 @@ func (s *Server) installWorkspaceScheduler(ctx context.Context, config *rest.Con
 	externalLogicalClusterAdminConfig = rest.CopyConfig(externalLogicalClusterAdminConfig)
 	externalLogicalClusterAdminConfig = rest.AddUserAgent(externalLogicalClusterAdminConfig, workspace.ControllerName+"+"+s.Options.Extra.ShardName)
 
-	workspaceController, err := workspace.NewController(
-		s.Options.Extra.ShardName,
-		kcpClusterClient,
-		kubeClusterClient,
-		logicalClusterAdminConfig,
-		externalLogicalClusterAdminConfig,
-		s.KcpSharedInformerFactory.Tenancy().V1alpha1().Workspaces(),
-		s.CacheKcpSharedInformerFactory.Core().V1alpha1().Shards(),
-		s.CacheKcpSharedInformerFactory.Tenancy().V1alpha1().WorkspaceTypes(),
-		s.KcpSharedInformerFactory.Core().V1alpha1().LogicalClusters(),
-	)
-	if err != nil {
-		return err
-	}
+	indexers.AddIfNotPresentOrDie(s.KcpSharedInformerFactory.Tenancy().V1alpha1().Workspaces().Informer().GetIndexer(), cache.Indexers{
+		unschedulable: indexUnschedulable,
+	})
+	indexers.AddIfNotPresentOrDie(s.CacheKcpSharedInformerFactory.Core().V1alpha1().Shards().Informer().GetIndexer(), cache.Indexers{
+		byBase36Sha224Name: indexByBase36Sha224Name,
+	})
+	indexers.AddIfNotPresentOrDie(s.CacheKcpSharedInformerFactory.Tenancy().V1alpha1().WorkspaceTypes().Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPathAndName: indexers.IndexByLogicalClusterPathAndName,
+	})
 
 	if err := s.registerController(&controllerWrapper{
 		Name: workspace.ControllerName,
 		Runner: func(ctx context.Context) {
+			workspaceController, err := workspace.NewController(
```

Comment:
this issue happens with every controller, we need to to do it for the rest of the controllers

Reply:
After this PR popped up I _was_ wondering why it would only affect specific controllers, this seemed like a fairly fundamental flaw in how I implemented leader election.

@sankar17 is this something that you want to refactor at a bigger scale to address the architectural problem or would you like me to take a look?

Reply:
@embik This has to be fixed for all the controllers, initially we got stuck with workspace controller creation stopped working with our longevity tests / over night test with 200 workspace creation / deletion . We saw the kcp metrics and the request queued to all 3/3 pods, though the leader election is enabled.

Currently I am working on changing this logic for all the controllers and need to do more tests.

if you think this is going to be fixed architectural level , please take a look at it.

CC: @yastij @palnabarun

Reply:
Okay, taking a quick look at where queues are being started I think the approach in this PR is overall fine. @sankar17 feel free to carry on, we should change it for all controllers. I can also take over if you don't have time.

Reply:
@embik I did changes to all the controllers, currently testing the fix, if all good I will update this PR in few days